minetest/minetest
# Add bulk ABMs

Would fix https://github.com/minetest/minetest/issues/11400 if merged. Usecases of this PR are mentioned there.

The PR adds "bulk ABMs" which execute up to once per ABM run with a list of node positions. A list of nodes is not provided, as this list may become out of date. By the time the Lua callback is called, the nodes may have changed, so the callback should do its own checking. To make an ABM run in bulk, put the action in the field `bulk_action` instead of `action`. The action will then be given a list of positions when it runs.

The PR adds a new abstract subclass of `ActiveBlockModifier` called `BulkActiveBlockModifier`. Another class `LuaBulkABM` was also added that implements `BulkActiveBlockModifier`. The functionality shared between `LuaABM` and `LuaBulkABM` has been put into a common parent class `AbstractLuaABM`.

## To do

This PR is ready for review.

- [x] A way to limit the time taken by bulk ABMs as with other ABMs.
- [x] `minetest.features` entry

## How to test

I have added a moving test node `testnodes:bulk_abm`.

You can also test with this mod: [vm_liquid.zip](https://github.com/minetest/minetest/files/9429489/vm_liquid.zip)


It's sort of like a simpler version of dynamic_liquid. It makes water sources flow downward and sideways into the available space. Because it uses a bulk ABM, it can make use of voxel manipulators. On my computer, the ABM's `bulk_action` takes about 20 milliseconds to run with about 10000 nodes. This does not include the time taken by the C++ code scanning for nodes and stuff.

To test the ABM time limit, place a huge area of water sources with World Edit, then let these flow until the ABM time starts hitting the limit. The time shouldn't exceed the limit by too much.

Also make sure regular ABMs still work.
